hermes/probe/monitor_test.go

```diff
@@ -0,0 +1,123 @@
+
+	metricpb "github.com/google/cloudprober/metrics/proto"
+	probes_configpb "github.com/google/cloudprober/probes/proto"
+	monitorpb "github.com/googleinterns/step224-2020/config/proto"
+)
+
+// GenTestConfig generates a test HermesProbeDef proto config for
+// initialising and running a MonitorProbe.
+// Arguments:
+//	- name: pass the name for this probe instance.
+// Returns:
+//	- probeDef: a test probe config with a HermesProbeDef extension set.
+//	- hermesExtension: returns the HermesProbeDef extension.
+func GenTestConfig(name string) (*probes_configpb.ProbeDef, *monitorpb.HermesProbeDef) {
+	probeDef := &probes_configpb.ProbeDef{
+		Name: proto.String(name),
+		Type: probes_configpb.ProbeDef_EXTENSION.Enum(),
+	}
+
+	hermesExtension := &monitorpb.HermesProbeDef{
+		ProbeName:    proto.String(name),
+		Targets:      []*monitorpb.Target{},
+		TargetSystem: monitorpb.HermesProbeDef_GCS.Enum(),
+		IntervalSec:  proto.Int32(3600),
+		TimeoutSec:   proto.Int32(60),
+		ProbeLatencyDistribution: &metricpb.Dist{
+			Buckets: &metricpb.Dist_ExplicitBuckets{
+				ExplicitBuckets: "0.1, 0.2, 0.4, 0.6 0.8, 1.6, 3.2, 6.4, 12.8, 1000",
+			},
+		},
+		ApiCallLatencyDistribution: &metricpb.Dist{
+			Buckets: &metricpb.Dist_ExplicitBuckets{
+				ExplicitBuckets: "0.000000002, 0.000000004, 0.000000008, 0.000000016, 0.00000032, 0.000000064, 0.000000128, 100",
+			},
+		},
+	}
+	proto.SetExtension(probeDef, monitorpb.E_HermesProbeDef_HermesProbeDef, hermesExtension)
+	return probeDef, hermesExtension
+}
+
+// GenOptsFromConfig generates the options for the MonitorProbe
+// from the probe config passed.
+// Arguments:
+//	- t: testing object for throwing errors
+//	- cfg: probe config.
+// Returns:
+//	- opts: options for a probe generated from the probe config.
+func GenOptsFromConfig(t *testing.T, cfg *monitorpb.HermesProbeDef) *options.Options {
+	opts := &options.Options{
+		Interval: time.Duration(cfg.GetIntervalSec()) * time.Second,
+		Timeout:  time.Duration(cfg.GetTimeoutSec()) * time.Second,
+	}
+
+	var err error
+	if opts.Logger, err = logger.NewCloudproberLog(cfg.GetProbeName()); err != nil {
+		t.Fatalf("error in initializing logger for the probe (%s): %v", cfg.GetProbeName(), err)
+	}
+
+	opts.ProbeConf = cfg
+
+	return opts
+}
+
+func TestInit(t *testing.T) {
+	wantName := "testProbe1"
+	mp := &MonitorProbe{}
+	_, wantConfig := GenTestConfig(wantName)
+	wantOpts := GenOptsFromConfig(t, wantConfig)
+
+	if err := mp.Init(wantName, wantOpts); err != nil {
+		t.Errorf("error when calling Init() on MonitorProbe object: wanted %v, got %v for error return value", nil, err)
+	}
+	if mp.name != wantName {
+		t.Errorf("error when calling Init() on MonitorProbe, wanted %v, got %v for name field", wantName, mp.name)
+	}
+	if !reflect.DeepEqual(mp.Config, wantConfig) {
+		t.Errorf("error when calling Init() on MonitorProbe, wanted %v, got %v for config field", wantConfig, mp.Config)
+	}
+	if !reflect.DeepEqual(mp.targets, wantConfig.GetTargets()) {
```

Comment:
Not that this is wrong to have this check here, but it's preferable to test the *behaviour* of the package via its public API.
Its entirely possible that the `targets` field is initialized to the expected value, but then it's not *used* in the expected manner.

The problem with inspecting non-exported fields is that these are usually implementation details.
Ideally tests don't have to be changed when an internal implementation detail changes.
Imagine changing how the list of targets is stored: for example switching from string slice to a map.
Now you have to change the test as well, which is just busy-work.

Reply:
Given that the Init() function only assigns values to other values and creates empty structs, would it be worthwhile removed the checks and just calling `Init()` and checking the error?

Reply:
I would do exactly that: call Init() and check there's no error.
In other tests, Init() will be part of the set-up. If those tests pass, there's implicit confirmation that Init() did its job.

Reply:
Excellent! Thank you, I'll go with this approach.